### ISSUE TITLE
PSA: HF transformers implementation open sourced (with Trainer support)

### ISSUE THREAD

**Issue by thomasgauthier** [SesameAILabs/csm#129]
(Re-posting from HF for visibility)

Hi everyone,

First I'd like to thank the Sesame team for this amazing model.

I want to share with the community something I've been working on: a re-implementation of the model for HuggingFace transformers, fully compatible with `Trainer`.

It supports decoder training amortization like presented in the [CSM blog post](https://www.sesame.com/research/crossing_the_uncanny_valley_of_voice#:~:text=The%20audio%20decoder%20is%20trained%20on%20only%20a%20random%201/16%20subset%20of%20the%20audio%20frames%2C%20while%20the%20zeroth%20codebook%20is%20trained%20on%20every%20frame.). It also supports generation.

It is Apache 2.0 licensed, the code can be found here: https://github.com/thomasgauthier/csm-hf
The converted pretrained model weights are hosted at https://huggingface.co/thomasgauthier/csm-1b-hf

Looking forward to see what the community will do with this.

🤞

**Comment by ZackHodari** [comment]
Great to see contributions for the community like these! Thanks

Have you tested fine-tuning yet? If so, I'm curious how well do the losses behave and how well does it learn from the data you tried?

**Comment by thomasgauthier** [comment]
Yes finetuning works for me. The model learns. Haven't had any particularly good models out of it yet (audio quality seems to drop a little bit after finetuning) but it does capture speaker features I care about. With a smallish dataset the model doesn't learn to copy the voice unconditionally, it still uses a random voice at generation time if just fed text. But when given some reference audio in-context as example, voice cloning works.

Here is what my loss curves look like:

![Image](https://github.com/user-attachments/assets/646014b2-3874-4c5b-9833-b1d84da7f5c2)

You can see it's pretty noisy, but it seems to converge.

**Comment by tarun7r** [comment]
@thomasgauthier Hey, really appreciate the great work — the finetuning does a solid job capturing speaker features. Quick question: Can we finetune the model for a new language? If yes, what would the key requirements be?
Thanks again!

**Comment by thomasgauthier** [comment]
@tarun7r Yes, you can finetune for a new language. Only limitation right now is that my code is not optimized for multi-GPU setup, so depending on dataset size you might be looking at very long training run. The Orpheus team claims [5k hours is enough](https://canopylabs.ai/releases/orpheus_can_speak_any_language#training) for adding a new language. I think you can train this on a single GPU within a few days. [They also suggest](https://reddit.com/r/LocalLLaMA/comments/1jo88lg/part_of_orpheus_team_here_ama_educational_content/mkxfyyx/) a learning rate of of 5e-5.
